lib/semantics/resolve-labels.cc

```diff
@@ -0,0 +1,917 @@
+    }
+    for (const auto &subpart : std::get<1>(a.t)) {
+      if (!PresentAndEq(std::get<std::optional<parser::Name>>(
+                            std::get<0>(subpart.t).statement.t),
+              constructName)) {
+        errorHandler_.Say(currentPosition_,
+            parser::MessageFormattedText{
+                "%sCASE statement name mismatch"_err_en_US, selectSubTag});
+      }
+    }
+  }
+
+  // C1033
+  void CheckName(const parser::WhereConstruct &whereConstruct) {
+    const auto &constructName{std::get<std::optional<parser::Name>>(
+        std::get<parser::Statement<parser::WhereConstructStmt>>(
+            whereConstruct.t)
+            .statement.t)};
+    if (!BothEqOrNone(constructName,
+            std::get<parser::Statement<parser::EndWhereStmt>>(whereConstruct.t)
+                .statement.v)) {
+      errorHandler_.Say(currentPosition_,
+          parser::MessageFormattedText{
+              "WHERE construct name mismatch"_err_en_US});
+    }
+    for (const auto &maskedElsewhere :
+        std::get<std::list<parser::WhereConstruct::MaskedElsewhere>>(
+            whereConstruct.t)) {
+      if (!PresentAndEq(
+              std::get<std::optional<parser::Name>>(
+                  std::get<parser::Statement<parser::MaskedElsewhereStmt>>(
+                      maskedElsewhere.t)
+                      .statement.t),
+              constructName)) {
+        errorHandler_.Say(currentPosition_,
+            parser::MessageFormattedText{
+                "ELSEWHERE (<mask>) statement name mismatch"_err_en_US});
+      }
+    }
+    if (std::get<std::optional<parser::WhereConstruct::Elsewhere>>(
+            whereConstruct.t)
+            .has_value()) {
+      if (!PresentAndEq(
+              std::get<parser::Statement<parser::ElsewhereStmt>>(
+                  std::get<std::optional<parser::WhereConstruct::Elsewhere>>(
+                      whereConstruct.t)
+                      ->t)
+                  .statement.v,
+              constructName)) {
+        errorHandler_.Say(currentPosition_,
+            parser::MessageFormattedText{
+                "ELSEWHERE statement name mismatch"_err_en_US});
+      }
+    }
+  }
+
+  // C1134, C1166
+  template<typename A>
+  void CheckLabelContext(const char *const stmtString, const A &constructName) {
+    const auto iter{std::find(
+        constructNames_.crbegin(), constructNames_.crend(), constructName)};
+    if (iter == constructNames_.crend()) {
+      errorHandler_.Say(currentPosition_,
+          parser::MessageFormattedText{
+              "%s construct-name '%s' is not in scope"_err_en_US, stmtString,
+              constructName.c_str()});
+    }
+  }
+
+  // 6.2.5, paragraph 2
+  void CheckLabelInRange(parser::Label label) {
+    if (label < 1 || label > 99999) {
+      errorHandler_.Say(currentPosition_,
+          parser::MessageFormattedText{
+              "label '%" PRIu64 "' is out of range"_err_en_US, label});
```

Comment:
Rather than using type-specific formats that need to be found and updated when/if the `parser::Label` type alias ever changes, consider using just `%d` for the format and `static_cast<int>(label)` for the argument.  That would be future-proof and maybe even more readable.